configparser.go

```diff
@@ -200,3 +191,82 @@ func (p *ConfigParser) SaveWithDelimiter(filename, delimiter string) error {
 
 	return nil
 }
+
+// ParseReader parses data into ConfigParser from provided reader.
+func (p *ConfigParser) ParseReader(in io.Reader) error {
+	reader := bufio.NewReader(in)
+	var lineNo int
+	var curSect *Section
+
+	keyValue := regexp.MustCompile(
+		fmt.Sprintf(
+			`([^%[1]s\s][^%[1]s]*)\s*(?P<vi>[%[1]s]+)\s*(.*)$`,
+			p.opt.delimiters,
+		),
+	)
+	keyWNoValue := regexp.MustCompile(
+		fmt.Sprintf(
+			`([^%[1]s\s][^%[1]s]*)\s*((?P<vi>[%[1]s]+)\s*(.*)$)?`,
+			p.opt.delimiters,
+		),
+	)
+
+	for {
+		l, _, err := reader.ReadLine()
+		if err != nil {
+			break
+		}
+		lineNo++
+		if len(l) == 0 {
+			continue
+		}
+		line := strings.TrimFunc(string(l), unicode.IsSpace) // ensures sectionHeader regex will match
+
+		// Skip comment lines and empty lines
+		if p.opt.commentPrefixes.AtFirst(line) || line == "" {
+			continue
+		}
+
+		if match := sectionHeader.FindStringSubmatch(line); len(match) > 0 {
+			section := p.opt.inlineCommentPrefixes.Split(match[1])
+			if section == p.opt.defaultSection {
+				curSect = p.defaults
+			} else if _, present := p.config[section]; !present {
+				curSect = newSection(section)
+				p.config[section] = curSect
+			} else if p.opt.strict {
+				return fmt.Errorf("section %q error: %w", section, ErrAlreadyExist)
```

Comment:
I'm not sure wrapping a sentinel error is the right way to do this.

I'd be inclined to define an `AlreadyExistsError` type, which can contain the section with `.Error() string` which would return this message.

Take a look at the stdlib [PathError](https://pkg.go.dev/io/fs#PathError) for an example.

Reply:
Hm, I also use it [here](https://github.com/emar-kar/go-configparser/blob/dc2303d71389a6fd6ce2d1deaa065613196516cd/methods.go#L292) to report duplicate option. As I see `PathError` still have error field and `Unwrap` method and `Op`. Is it worth it to make the whole new structure for the error which actually can be triggered twice? Imho wrapped error is more informative, but at the same time can be easily unwrapped in the user code to assert it with `errors.Is`

Reply:
I've removed wrapped error and let it be just text description, without any specific type